Add Blend Import

fire opened the pull request:
Fixes https://github.com/godotengine/godot-proposals/issues/3529

Discussed with the Godot Discord community and with Reduz.

lyuma commented:
Just chatted with @fire and I think it would make more sense to move the blender path into EditorSettings, since I do not want to input my blender executable path in every single project on windows, and also would not want to check it into version control by mistake.

I am also apprehensive about the settings for the blender python scripts being per-project. There was an idea of some use-case where a user customizes the blender export to hack in additional features, but my opinion is that level of customization would be best left to [custom GDScript addons](https://github.com/V-Sekai/godot-blender).

Hardcoding the python strings would be easiest. I'd also be ok with having the scripts as an editor setting.

fire commented:
QA tested reports

1. skeletons sometimes loses mesh
2. exporting godot uses the default cube, light and camera.

fire commented:
Filled in the description xml for EditorSceneFormatImporterBlend.

fire commented:
Todo: add an explicit enable per project.

fire commented:
Due to non-consensus, I'll move export to a separate branch. https://github.com/V-Sekai/godot/tree/blend-export

reduz commented on modules/gltf_blend/register_types.cpp:
```diff
@@ -0,0 +1,73 @@
+static void _editor_init() {
+	_GLOBAL_DEF("filesystem/import/blend/enabled", true, true);
+	bool blender_enabled =
+			EditorSettings::get_singleton()->get("filesystem/import/blend/enabled");
```
This should be a project setting, not an editor setting.

fire replied:
_GLOBAL_DEF macro calls project settings. The second part is a typo. Resolved.

rcorre commented on modules/gltf_blend/editor_scene_importer_blend.cpp:
```diff
@@ -0,0 +1,113 @@
+									"export_texture_dir='GODOT_TEXTURE_DIR',"
+									"export_colors=True,"
+									"export_all_influences=False,"
+									"export_extras=True,"
```
It would be neat to customize these parameters as scene import options. For example, I often use modifiers (especially `mirror`) so sometimes I'd want `export_apply=True`. However, that prevents exporting shapekeys, so I might want `export_apply=False` on other models (https://docs.blender.org/api/current/bpy.ops.export_scene.html).

I wasn't able to do that when working on a plugin, as it seems like `EditorSceneImporter` can't add custom options like `ResourceImporter` can, so I don't know if it is possible at the moment. If not, would folks be open to a proposal giving `EditorSceneImporter` an equivalent of `EditorImportPlugin.get_import_options`?

rcorre replied:
Or, since `export_apply` is the only one I can think of to customize at the moment, maybe the logic could be:

- If there are modifiers but no shapekeys, set `export_apply=True`
- If there are shapekeys but no modifiers, set `export_apply=False`
- If there are shapekeys and modifiers, raise a warning

That probably requires more python.

fire replied:
Reduz said he's working on this after this pr is merged.

rcorre commented on modules/gltf_blend/editor_scene_importer_blend.cpp:
```diff
@@ -0,0 +1,113 @@
+		print_line(standard_out[line_i]);
+	}
+	if (ret != OK) {
+		print_error("Blender returned " + itos(ret));
```
Do we need to set `r_err` here? 

When testing the plugin, I was using it with a CI build system, which I had forgotten to install blender on. The export printed errors, but since it returned `0` the build system reported a success, and I never knew to look for the errors.

There doesn't seem to be a way for a `gdscript` plugin to report an error, but I'm wondering if `r_err` will do that here?

fire replied:
I'll set r_err.

fire commented:
I'll leave a message with @aaronfranke and @neikeq asking if they have an idea how to resolve the cicd error.

raulsntos commented on modules/gltf_blend/register_types.cpp:
```diff
@@ -0,0 +1,74 @@
+void register_gltf_blend_types() {
+#ifndef _3D_DISABLED
+#ifdef TOOLS_ENABLED
+	ClassDB::APIType prev_api = ClassDB::get_current_api();
+	ClassDB::set_current_api(ClassDB::API_EDITOR);
+	ClassDB::set_current_api(prev_api);
+	GDREGISTER_CLASS(EditorSceneFormatImporterBlend);
+	EditorNode::add_init_callback(_editor_init);
```
I think this is supposed to be like this:

```suggestion
	ClassDB::APIType prev_api = ClassDB::get_current_api();
	ClassDB::set_current_api(ClassDB::API_EDITOR);
	GDREGISTER_CLASS(EditorSceneFormatImporterBlend);
	ClassDB::set_current_api(prev_api);
	EditorNode::add_init_callback(_editor_init);
```

neikeq replied:
Was only the C# CI failing? I thought I added some checks for this to the ClassDB tests.

rcorre commented on modules/gltf_blend/editor_scene_importer_blend.cpp:
```diff
@@ -0,0 +1,116 @@
+	Array standard_out;
+	int32_t ret = core_bind::OS::get_singleton()->execute(addon_path, args, standard_out, true);
+	for (int32_t line_i = 0; line_i < standard_out.size(); line_i++) {
+		print_line(standard_out[line_i]);
```
Do we still want to leave the (non-error)`print`s in? It might be a bit noisy compared to other importers. Here's a sample:

```
Godot Engine v3.4.stable.arch_linux (c) 2007-2021 Juan Linietsky, Ariel Manzur & Godot Contributors.
--- GDScript language server started ---
[--background, --python-expr, import bpy, os, sys;bpy.ops.wm.open_mainfile(filepath='/home/rcorre/src/godot-blender/example.blend');bpy.ops.export_scene.gltf(filepath='/home/rcorre/src/godot-blender/.import/example.blend-1ba6972ae2a764efa4ba948c0919249c.glb',export_format='GLB',export_colors=True,export_all_influences=False,export_extras=True,export_cameras=True,export_lights=True);]
'/usr/bin/2.93/python/lib/python3.9/site-packages/libextern_draco.so' does not exist, draco mesh compression not available
11:59:01 | INFO: Starting glTF 2.0 export
11:59:01 | INFO: Extracting primitive: Cube
11:59:01 | INFO: Primitives created: 1
11:59:01 | INFO: Finished glTF 2.0 export in 0.0035049915313720703 s

Blender 2.93.5 (hash a791bdabd0b2 built 2021-11-06 11:01:22)
Read prefs: /home/rcorre/.config/blender/2.93/config/userpref.blend
/run/user/1000/gvfs/ non-existent directory
Read blend: /home/rcorre/src/godot-blender/example.blend

Blender quit
```

fire replied:
Debugging blender was difficult without this.

akien-mga replied:
Well once it's working, this would still be too verbose. If it's only for debugging, it should be `print_verbose` at the very least.

rcorre commented on modules/gltf_blend/register_types.cpp:
```diff
@@ -0,0 +1,74 @@
+	if (blender_path.is_empty()) {
+		return;
+	}
+	Ref<EditorSceneFormatImporterBlend> import_gltf;
```
nit: should the variable be named `import_blend`?

fire replied:
Will rename.

reduz commented:
Before merging, please wait until #54976 is merged.

reduz commented:
Since now you can add options for blend files, I suggest the following:
* Always export GLTF Separate with the "Keep Original" setting on (do not let user choose).
* Custom Properties, Cameras and Punctual lights should be exported by default (on by default)
* Limit to should probably limit to visible objects by default (what does Unity do here?)
* +Y-Up should be always on and **not exposed** to the user.
* UVs, Normals and Tangents should be exposed, but **on by default**.
* Materials and Images setting should **not be exposed** (use Default and Automatic).
* Compression should **not be exposed** because we do away with the GLTF file anyway.
* Animations options all exposed, **on by default**.
* Shape Keys **not exposed**. Shape key tangets **always on**
* Skinning exposed by **Include all bone influences** on by default.

Additionally, It would be ideal that we use the the code from the [ESCN exporter](https://github.com/godotengine/godot-blender-exporter) and hook it into the Blender GLTF2 export hooks that converts Blender materials to Godot shaders, and we expose it as an option "Export materials as Godot shaders". This can be done at a later stage, though.

fire commented:
List of options. Please help craft the wording.

```json
{
"cameras": true,
"custom_properties": true,
"punctual_lights": true,
"export": { EXPORT_VISIBLE, EXPORT_ALL },
"uvs": true,
"normals": true,
"tangents": true,
"animation/limit_playback_range": true,
"animation/always_sample_animations": true,
"animation/group_tracks": true,
"animation/export_bones_deforming_mesh_only": true,
"skins": { SKIN_COMPATIBLE_INFLUENCES, SKIN_ALL_BONE_INFLUENCES },
}
```

rcorre commented:
I'd really love the `apply_modifiers` option to be exposed as well, though it might contradict with @reduz's request for:

> Shape Keys not exposed (always export if they exist). Shape key tangets always on

Would you be open to one of the following options?

1. Expose options for `export_modifiers` and `export_shapekeys` only allow one to be selected
2. Expose options for `export_modifiers` and `export_shapekeys`, warn if both are selected
3. Write some extra `python` to detect if shapekeys are present, and automatically set the options appropriately (maybe warning if you have both shapekeys and unapplied modifiers)

fire commented:
If you want to clean this up? https://github.com/V-Sekai/V-Sekai-Blender-tools/tree/main/addons/ApplyModifierForObjectWithShapeKeys

rcorre commented:
> I think getting apply modifers to work with blend shapes is out of scope for this pr.

Oh yeah, I wasn't suggesting that at all. `ApplyModifierForObjectWithShapeKeys` looks neat, but I'd be perfectly happy with the option to apply modifiers on models that _don't_ use shapekeys (which is most models, for me).

Calinou commented:
There's also the question of triangulation: should we perform it automatically on export? In general, triangulating a mesh before exporting it is recommended to ensure the triangle layout is deterministic across applications. This is especially important if you're using 3D painting software on the same mesh.

rcorre commented:
>  I remember some Blender exporters featuring a built-in triangulate option which doesn't require you to add a modifier

https://github.com/godotengine/godot-blender-exporter/blob/c1136facd90c3f1336c86e7a9921de889e46af5b/io_scene_godot/converters/utils.py#L35-L46 for reference

reduz commented on editor/import/editor_import_collada.cpp:
```diff
@@ -1755,7 +1755,7 @@ void EditorSceneFormatImporterCollada::get_extensions(List<String> *r_extensions
 }
 
-Node *EditorSceneFormatImporterCollada::import_scene(const String &p_path, uint32_t p_flags, int p_bake_fps, List<String> *r_missing_deps, Error *r_err) {
+Node *EditorSceneFormatImporterCollada::import_scene(const String &p_path, uint32_t p_flags, const Dictionary &p_options, int p_bake_fps, List<String> *r_missing_deps, Error *r_err) {
```
I would probably still pass the regular options map here, I suggest using the dictionary only for the function you expose for binding to script via GDVIRTUAL.

fire replied:
I am having trouble with this. Can you assist?

reduz replied:
Yes, I Suggest just pass `const Map<StringName, Variant> &p_options` in the C++ side, in the bindings side do:
```C++
	GDVIRTUAL4R(Object *, _import_scene, String, Dictionary, uint32_t, uint32_t)
	GDVIRTUAL4R(Ref<Animation>, _import_animation, Dictionary, String, uint32_t, uint32_t)
```

and pass a conversion of the options to a dictionary. You can make a small helper static function to use in both of those when doing the GDVIRTUAL_CALL.

fire replied:
Updated.

reduz commented on modules/gltf_blend/editor_scene_importer_blend.cpp:
```diff
@@ -0,0 +1,228 @@
+			"filesystem/blend/blender_path");
+	String json_parameters =
+			"{\"source\": \"GODOT_SOURCE\", \"sink\": \"GODOT_SINK\"}";
+	String texture_dir_string = p_options["blend/materials/texture_directory"];
```
I think this is a mistake, if you use the keep textures option, whathever you export will still reference the proper file on disk, so you should not bother about copying textures or anything like this.

fire replied:
Resolved in the next push.

reduz commented on modules/gltf_blend/editor_scene_importer_blend.cpp:
```diff
@@ -0,0 +1,228 @@
+	r_options->push_back(ResourceImporter::ImportOption(
+			PropertyInfo(Variant::BOOL, "blend/nodes/custom_properties"), true));
+	r_options->push_back(ResourceImporter::ImportOption(
+			PropertyInfo(Variant::INT, "blend/nodes/modifiers",
```
would use blender/ instead of blend/, feels less confusing

fire replied:
Will be in the next update.

fire commented:
While I wait for cicd in ~40 minutes, I'll do some QA.

akien-mga commented on modules/gltf_blend/SCsub:
```diff
@@ -0,0 +1,10 @@
+Import("env_modules")
+
+env_blend = env_modules.Clone()
+env_blend.Prepend(CPPPATH=["."])
```
This shouldn't be needed?

akien-mga commented on modules/gltf_blend/doc_classes/EditorSceneFormatImporterBlend.xml:
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="EditorSceneFormatImporterBlend" inherits="EditorSceneFormatImporter" version="4.0">
+	<brief_description>
+		EditorSceneFormatImporterBlend imports ".blend" files.
```
```suggestion
		Importer for Blender's [code].blend[/code] scene file format.
```

akien-mga commented on modules/gltf_blend/doc_classes/EditorSceneFormatImporterBlend.xml:
```diff
@@ -0,0 +1,17 @@
+		EditorSceneFormatImporterBlend imports ".blend" files.
+	</brief_description>
+	<description>
+		EditorSceneFormatImporterBlend imports ".blend" files through the glTF2 3d import pipeline.
+
+		The "filesystem/blender/blender_path" editor setting sets the location of the blender binary.
+
+		The "filesystem/blender/enabled" toogle can be found in the project setting.
+
+		Blender export using glTF2 "use keep separate" mode requires Blender 3.0 and above to reference external textures.
```
No empty lines in the XML.
```suggestion
		Imports Blender scenes in the [code].blend[/code] file format through the glTF 2.0 3D import pipeline. This importer requires Blender to be installed by the user, so that it can be used to export the scene as glTF 2.0.
		The location of the Blender binary is set via [member EditorSettings.filesystem/blender/blender_path].
		This importer is only used if [member ProjectSettings.filesystem/blender/enabled] is enabled, otherwise [code].blend[/code] present in the project folder are not imported.
		Blender export using glTF 2.0 "use keep separate" mode requires Blender 3.0 and above to reference external textures.
```
Please check the exact words and case style of "use keep separate" in the Blender 3.0 glTF export settings to use the same here.

akien-mga commented on modules/gltf_blend/config.py:
```diff
@@ -0,0 +1,16 @@
+def can_build(env, platform):
+    return not env["disable_3d"]
```
This would also need to check that the `gltf` module is enabled.

But isn't it a bit overkill to have a dedicated module for this? It could simply be a file in the `gltf` module, since it depends on the `gltf` importer.

akien-mga replied:
If kept as a module, it should also be excluded if `tools` is not enabled.

rcorre replied:
> If kept as a separate module, it should also be excluded if tools is not enabled.

Wouldn't that mean that a headless build used in a CI/CD pipeline wouldn't be able to export projects which import blender files?

akien-mga replied:
Headless builds are `tools=yes`, otherwise they can't export at all.

rcorre replied:
Ah, all good then

akien-mga commented on modules/gltf_blend/register_types.cpp:
```diff
@@ -0,0 +1,74 @@
+	if (!blender_enabled) {
+		return;
+	}
+	_EDITOR_DEF("filesystem/blender/blender_path", "blender", true);
```
This should be "filesystem/**import/**blender".

akien-mga replied:
We have clear macros without `_` private prefix and boolean trap, you should use them :)
```suggestion
	EDITOR_DEF_RST("filesystem/import/blender/blender_path", "blender");
```

akien-mga commented on modules/gltf_blend/register_types.cpp:
```diff
@@ -0,0 +1,74 @@
+#ifdef TOOLS_ENABLED
+static void _editor_init() {
+	_GLOBAL_DEF("filesystem/import/blender/enabled", false, true);
+	bool blender_enabled =
+			ProjectSettings::get_singleton()->get("filesystem/import/blender/enabled");
```
```suggestion
	bool blender_enabled = GLOBAL_GET("filesystem/import/blender/enabled");
```

akien-mga commented on modules/gltf_blend/register_types.cpp:
```diff
@@ -0,0 +1,74 @@
+	EditorSettings::get_singleton()->add_property_hint(
+			PropertyInfo(Variant::STRING, "filesystem/blender/blender_path",
+					PROPERTY_HINT_GLOBAL_FILE));
+	String blender_path =
+			EditorSettings::get_singleton()->get("filesystem/blender/blender_path");
```
```suggestion
	String blender_path = EDITOR_GET("filesystem/blender/blender_path");
```

akien-mga commented on modules/gltf_blend/editor_scene_importer_blend.h:
```diff
@@ -0,0 +1,75 @@
+	virtual Ref<Animation> import_animation(const String &p_path,
+			uint32_t p_flags, const Map<StringName, Variant> &p_options,
+			int p_bake_fps) override;
+	virtual void
+	get_import_options(const String &p_path,
+			List<ResourceImporter::ImportOption> *r_options) override;
+	virtual Variant
+	get_option_visibility(const String &p_path, const String &p_option,
+			const Map<StringName, Variant> &p_options) override;
```
That looks pretty weird. Make sure your IDE isn't configured to force hard wrapping on ridiculously low values.
```suggestion
	virtual void get_import_options(const String &p_path,
			List<ResourceImporter::ImportOption> *r_options) override;
	virtual Variant get_option_visibility(const String &p_path, const String &p_option,
			const Map<StringName, Variant> &p_options) override;
```

akien-mga commented on modules/gltf_blend/editor_scene_importer_blend.h:
```diff
@@ -0,0 +1,75 @@
+		BLEND_BONE_INFLUENCES_COMPATIBLE,
+		BLEND_BONE_INFLUENCES_ALL
+	};
+	enum { BLEND_MODIFIERS_NONE,
+		BLEND_MODIFIERS_ALL };
```
```suggestion
	enum {
		BLEND_MODIFIERS_NONE,
		BLEND_MODIFIERS_ALL
	};
```

akien-mga commented on modules/gltf_blend/editor_scene_importer_blend.h:
```diff
@@ -0,0 +1,75 @@
+#include "editor/editor_settings.h"
+#include "editor/import/resource_importer_scene.h"
+#include "scene/main/node.h"
+#include "scene/resources/packed_scene.h"
```
I don't see PackedScene used here.

fire replied:
Removed.

akien-mga commented on modules/gltf_blend/editor_scene_importer_blend.h:
```diff
@@ -0,0 +1,75 @@
+#include "core/variant/variant.h"
+#ifdef TOOLS_ENABLED
+
+#include "editor/editor_settings.h"
```
What uses EditorSettings in this header?

fire replied:
Removed.

akien-mga commented on modules/gltf_blend/editor_scene_importer_blend.h:
```diff
@@ -0,0 +1,75 @@
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef EDITOR_SCENE_IMPORTER_BLEND_H
+#define EDITOR_SCENE_IMPORTER_BLEND_H
+#include "core/variant/variant.h"
+#ifdef TOOLS_ENABLED
+
+#include "editor/editor_settings.h"
```
There's no need to define or include anything if not in tools.
```suggestion
#ifdef TOOLS_ENABLED

#ifndef EDITOR_SCENE_IMPORTER_BLEND_H
#define EDITOR_SCENE_IMPORTER_BLEND_H

#include "core/variant/variant.h"
#include "editor/editor_settings.h"
```

akien-mga commented on modules/gltf_blend/editor_scene_importer_blend.cpp:
```diff
@@ -0,0 +1,260 @@
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "core/config/project_settings.h"
+#include "core/error/error_list.h"
+#include "core/error/error_macros.h"
+#include "core/io/json.h"
+#include "core/object/object.h"
+#include "core/variant/dictionary.h"
+#include "modules/gltf/gltf_document.h"
+#if TOOLS_ENABLED
+#include "editor_scene_importer_blend.h"
+
+#include "core/core_bind.h"
+#include "core/object/ref_counted.h"
+#include "modules/gltf/editor_scene_importer_gltf.h"
+#include "scene/3d/node_3d.h"
+#include "scene/animation/animation_player.h"
+#include "scene/resources/animation.h"
```
Please tidy your includes to follow the style guide: https://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html#header-includes

And make sure that all the files you include are actually needed. You tend to copy-paste the same chunk of includes in all importers and I always have to clean up after you because half the includes are not necessary.

For example what uses `core_bind.h` here? That's a red flag, `core_bind.h` is to bind specific core functions for scripting langauges, you don't need that here, you have full access to the core functions already.

lyuma commented:
I'm not sure if the Map can be exposed to GDScript. I would suggest using `Dictionary` instead of `Map<StringName, Variant>`.

fire commented:
Please see the awful conversion code here. https://github.com/godotengine/godot/pull/54886/files#diff-849438ec58c77afcf1e87ce436d91f8dac1325c98ae9c9d6488707d0ca055517R78-R84 @lyuma for Dictionary to Map.

lyuma commented on modules/gltf_blend/doc_classes/EditorSceneFormatImporterBlend.xml:
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="EditorSceneFormatImporterBlend" inherits="EditorSceneFormatImporter" version="4.0">
+	<brief_description>
+		EditorSceneFormatImporterBlend imports ".blend" files.
+	</brief_description>
```
```suggestion
	<brief_description>
		Importer for Blender's [code].blend[/code] scene file format.
	</brief_description>
```

lyuma commented on modules/gltf_blend/doc_classes/EditorSceneFormatImporterBlend.xml:
```diff
@@ -0,0 +1,15 @@
+		The location of the Blender binary is set via editor settings "filesystem/blender/blender_path".
+		This importer is only used if project settings "filesystem/blender/enabled" is enabled, otherwise [code].blend[/code] present in the project folder are not imported.
+		Blend import requires Blender 3.0.
+		Internally, the EditorSceneFormatImporterBlend uses the Blender glTF "Use Original" mode to reference external textures. The "Use Original" mode uses the separate glTF 2.0 (.gltf + .bin + textures) Blender export.
```
```suggestion
		Internally, the EditorSceneFormatImporterBlend uses the Blender glTF "Use Original" mode to reference external textures.
```

lyuma commented on editor/import/resource_importer_scene.cpp:
```diff
@@ -101,25 +111,11 @@ Variant EditorSceneFormatImporter::get_option_visibility(const String &p_path, c
-}
-
 void EditorSceneFormatImporter::_bind_methods() {
-	ClassDB::bind_method(D_METHOD("import_scene_from_other_importer", "path", "flags", "bake_fps"), &EditorSceneFormatImporter::import_scene_from_other_importer);
-	ClassDB::bind_method(D_METHOD("import_animation_from_other_importer", "path", "flags", "bake_fps"), &EditorSceneFormatImporter::import_animation_from_other_importer);
-
```
why was this removed?

without this function, how can a user manually run the importer from GDScript?

fire replied:
The original use-case of an importer calling an external conversion helper (like, fbx2gltf) doesn't make sense when you can call them directly. 

```gdscript
var gltf = EditorSceneFormatImporterGLTF.new() 
gltf._import_scene() 
var fbx = EditorSceneFormatImporterFBX.new() 
fbx._import_scene() 
```

import_scene_from_other_importer and import_animation_from_other_importer apis are removed.

lyuma commented:
Fire explained that reduz specifically requested Map<StringName, Variant> instead of Dictionary, so you can ignore my comment about this. I see the conversion code in three places for the virtual 

Also, @fire noted that `import_scene_from_other_importer` was removed because `_import_scene` could be called from GDScript instead, for example manually invoking the FBX importer can be done with:
```gd
var fbx = EditorSceneFormatImporterFBX.new() 
fbx._import_scene()
```

lyuma commented:
Unfortunately, `_import_scene` is virtual and cannot be called.

I attempted the above code snippet but got this error instead:
```
 Invalid call. Nonexistent function '_import_scene' in base 'EditorSceneFormatImporterFBX'.
```

So I think we need another API or some other way to invoke the importer of our choosing.

lyuma commented:
Importing .blend files causes heap corruption on Windows. It is reproducible 100% of the time on my computer on every blend file. I have not had any successfully import yet.
```
The thread 0x12658 has exited with code 0 (0x0).
Critical error detected c0000374
godot.windows.tools.64.exe has triggered a breakpoint.
```

Run 1 stacktrace (where the error is detected, after it happened):
```
 	ntdll.dll!RtlReportCriticalFailure()	Unknown
 	ntdll.dll!RtlpHeapHandleError()	Unknown
 	ntdll.dll!RtlpHpHeapHandleError()	Unknown
 	ntdll.dll!RtlpLogHeapFailure()	Unknown
 	ntdll.dll!RtlpAnalyzeHeapFailure()	Unknown
 	ntdll.dll!RtlpFreeHeap()	Unknown
 	ntdll.dll!RtlpFreeHeapInternal()	Unknown
 	ntdll.dll!RtlFreeHeap()	Unknown
 	godot.windows.tools.64.exe!_free_base(void * block) Line 105	C++
 	godot.windows.tools.64.exe!Memory::free_static(void * p_ptr, bool p_pad_align) Line 169	C++
>	godot.windows.tools.64.exe!LocalVector<Vector3,unsigned int,0>::reset() Line 115	C++
 	godot.windows.tools.64.exe!LocalVector<Vector3,unsigned int,0>::~LocalVector<Vector3,unsigned int,0>() Line 254	C++
 	godot.windows.tools.64.exe!ImporterMesh::generate_lods(float p_normal_merge_angle, float p_normal_split_angle) Line 613	C++
 	godot.windows.tools.64.exe!ResourceImporterScene::_generate_meshes(Node * p_node, const Dictionary & p_mesh_data, bool p_generate_lods, bool p_create_shadow_meshes, ResourceImporterScene::LightBakeMode p_light_bake_mode, float p_lightmap_texel_size, const Vector<unsigned char> & p_src_lightmap_cache, Vector<Vector<unsigned char> > & r_lightmap_caches) Line 1612	C++
 	godot.windows.tools.64.exe!ResourceImporterScene::_generate_meshes(Node * p_node, const Dictionary & p_mesh_data, bool p_generate_lods, bool p_create_shadow_meshes, ResourceImporterScene::LightBakeMode p_light_bake_mode, float p_lightmap_texel_size, const Vector<unsigned char> & p_src_lightmap_cache, Vector<Vector<unsigned char> > & r_lightmap_caches) Line 1698	C++
 	godot.windows.tools.64.exe!ResourceImporterScene::_generate_meshes(Node * p_node, const Dictionary & p_mesh_data, bool p_generate_lods, bool p_create_shadow_meshes, ResourceImporterScene::LightBakeMode p_light_bake_mode, float p_lightmap_texel_size, const Vector<unsigned char> & p_src_lightmap_cache, Vector<Vector<unsigned char> > & r_lightmap_caches) Line 1698	C++
 	godot.windows.tools.64.exe!ResourceImporterScene::import(const String & p_source_file, const String & p_save_path, const Map<StringName,Variant,Comparator<StringName>,DefaultAllocator> & p_options, List<String,DefaultAllocator> * r_platform_variants, List<String,DefaultAllocator> * r_gen_files, Variant * r_metadata) Line 2036	C++
 	godot.windows.tools.64.exe!EditorFileSystem::_reimport_file(const String & p_file, const Map<StringName,Variant,Comparator<StringName>,DefaultAllocator> * p_custom_options, const String & p_custom_importer) Line 1901	C++
 	godot.windows.tools.64.exe!EditorFileSystem::reimport_files(const Vector<String> & p_files) Line 2151	C++
 	godot.windows.tools.64.exe!EditorFileSystem::_update_scan_actions() Line 627	C++
 	godot.windows.tools.64.exe!EditorFileSystem::_notification(int p_what) Line 1224	C++
```

and run 2 stacktrace:
```
 	ntdll.dll!RtlReportCriticalFailure()	Unknown
 	ntdll.dll!RtlpHeapHandleError()	Unknown
 	ntdll.dll!RtlpHpHeapHandleError()	Unknown
 	ntdll.dll!RtlpLogHeapFailure()	Unknown
 	ntdll.dll!RtlpAllocateHeap()	Unknown
 	ntdll.dll!RtlpAllocateHeapInternal()	Unknown
 	ntdll.dll!RtlpAllocateUserBlockFromHeap()	Unknown
 	ntdll.dll!RtlpAllocateUserBlock()	Unknown
 	ntdll.dll!RtlpLowFragHeapAllocFromContext()	Unknown
 	ntdll.dll!RtlpAllocateHeapInternal()	Unknown
 	godot.windows.tools.64.exe!_malloc_base(unsigned __int64 size) Line 34	C++
 	[Inline Frame] godot.windows.tools.64.exe!_aligned_offset_malloc_base(unsigned __int64 align, unsigned __int64) Line 169	C++
 	[Inline Frame] godot.windows.tools.64.exe!_aligned_malloc_base(unsigned __int64) Line 122	C++
 	godot.windows.tools.64.exe!_aligned_malloc(unsigned __int64 size, unsigned __int64 alignment) Line 555	C++
>	godot.windows.tools.64.exe!embree::alignedMalloc(unsigned __int64 size, unsigned __int64 align) Line 23	C++
 	[Inline Frame] godot.windows.tools.64.exe!embree::Scene::operator new(unsigned __int64) Line 29	C++
 	godot.windows.tools.64.exe!rtcNewScene(RTCDeviceTy * hdevice) Line 163	C++
 	godot.windows.tools.64.exe!StaticRaycasterEmbree::StaticRaycasterEmbree() Line 128	C++
 	godot.windows.tools.64.exe!StaticRaycasterEmbree::create_embree_raycaster() Line 42	C++
 	godot.windows.tools.64.exe!StaticRaycaster::create() Line 37	C++
 	godot.windows.tools.64.exe!ImporterMesh::generate_lods(float p_normal_merge_angle, float p_normal_split_angle) Line 401	C++
 	godot.windows.tools.64.exe!ResourceImporterScene::_generate_meshes(Node * p_node, const Dictionary & p_mesh_data, bool p_generate_lods, bool p_create_shadow_meshes, ResourceImporterScene::LightBakeMode p_light_bake_mode, float p_lightmap_texel_size, const Vector<unsigned char> & p_src_lightmap_cache, Vector<Vector<unsigned char> > & r_lightmap_caches) Line 1612	C++
 	godot.windows.tools.64.exe!ResourceImporterScene::_generate_meshes(Node * p_node, const Dictionary & p_mesh_data, bool p_generate_lods, bool p_create_shadow_meshes, ResourceImporterScene::LightBakeMode p_light_bake_mode, float p_lightmap_texel_size, const Vector<unsigned char> & p_src_lightmap_cache, Vector<Vector<unsigned char> > & r_lightmap_caches) Line 1698	C++
 	godot.windows.tools.64.exe!ResourceImporterScene::_generate_meshes(Node * p_node, const Dictionary & p_mesh_data, bool p_generate_lods, bool p_create_shadow_meshes, ResourceImporterScene::LightBakeMode p_light_bake_mode, float p_lightmap_texel_size, const Vector<unsigned char> & p_src_lightmap_cache, Vector<Vector<unsigned char> > & r_lightmap_caches) Line 1698	C++
 	godot.windows.tools.64.exe!ResourceImporterScene::import(const String & p_source_file, const String & p_save_path, const Map<StringName,Variant,Comparator<StringName>,DefaultAllocator> & p_options, List<String,DefaultAllocator> * r_platform_variants, List<String,DefaultAllocator> * r_gen_files, Variant * r_metadata) Line 2036	C++
 	godot.windows.tools.64.exe!EditorFileSystem::_reimport_file(const String & p_file, const Map<StringName,Variant,Comparator<StringName>,DefaultAllocator> * p_custom_options, const String & p_custom_importer) Line 1901	C++
 	godot.windows.tools.64.exe!EditorFileSystem::reimport_files(const Vector<String> & p_files) Line 2151	C++
```
are two example stacktraces of when the error was **DETECTED**. I have no information on when the heap corruption **OCCURRED**. Please take these stacktraces with a dose of salt.

Crash 3: Did not crash in the allocator, but similar issue
```
Exception thrown at 0x00007FF66F755F0D in godot.windows.tools.64.exe: 0xC0000005: Access violation writing location 0x0000021FD3A03000.

>	godot.windows.tools.64.exe!ImporterMesh::generate_lods(float p_normal_merge_angle, float p_normal_split_angle) Line 299	C++
 	godot.windows.tools.64.exe!ResourceImporterScene::_generate_meshes(Node * p_node, const Dictionary & p_mesh_data, bool p_generate_lods, bool p_create_shadow_meshes, ResourceImporterScene::LightBakeMode p_light_bake_mode, float p_lightmap_texel_size, const Vector<unsigned char> & p_src_lightmap_cache, Vector<Vector<unsigned char> > & r_lightmap_caches) Line 1612	C++
 	godot.windows.tools.64.exe!ResourceImporterScene::_generate_meshes(Node * p_node, const Dictionary & p_mesh_data, bool p_generate_lods, bool p_create_shadow_meshes, ResourceImporterScene::LightBakeMode p_light_bake_mode, float p_lightmap_texel_size, const Vector<unsigned char> & p_src_lightmap_cache, Vector<Vector<unsigned char> > & r_lightmap_caches) Line 1698	C++
 	godot.windows.tools.64.exe!ResourceImporterScene::_generate_meshes(Node * p_node, const Dictionary & p_mesh_data, bool p_generate_lods, bool p_create_shadow_meshes, ResourceImporterScene::LightBakeMode p_light_bake_mode, float p_lightmap_texel_size, const Vector<unsigned char> & p_src_lightmap_cache, Vector<Vector<unsigned char> > & r_lightmap_caches) Line 1698	C++
 	godot.windows.tools.64.exe!ResourceImporterScene::import(const String & p_source_file, const String & p_save_path, const Map<StringName,Variant,Comparator<StringName>,DefaultAllocator> & p_options, List<String,DefaultAllocator> * r_platform_variants, List<String,DefaultAllocator> * r_gen_files, Variant * r_metadata) Line 2036	C++
 	godot.windows.tools.64.exe!EditorFileSystem::_reimport_file(const String & p_file, const Map<StringName,Variant,Comparator<StringName>,DefaultAllocator> * p_custom_options, const String & p_custom_importer) Line 1901	C++
 	godot.windows.tools.64.exe!EditorFileSystem::reimport_files(const Vector<String> & p_files) Line 2151	C++
 	godot.windows.tools.64.exe!EditorFileSystem::_update_scan_actions() Line 627	C++
```

Also, it seems to be using the absolute path of the texture, but putting res in front.

For example, a texture located at `S:\Projects\assets\blender 1\Textures\Body Texture.png` will cause godot to load
`res://Projects/assets/blender 1/Textures/Body Texture.png`
this error:
```
ERROR: Resource file not found: res://Projects/assets/blender 1/Textures/Body Texture.png.
   at: (core\io\resource_loader.cpp:210)
ERROR: Condition "array_len == 0" is true. Returning: ERR_INVALID_DATA
   at: RenderingServer::mesh_create_surface_data_from_arrays (servers\rendering_server.cpp:903)
```


Additionally, it is causing GLTF errors when importing blend files:
```
ERROR: Invalid image definition in glTF file, it should specific an 'uri' or 'bufferView'.
   at: (modules\gltf\gltf_document.cpp:3032)
ERROR: Invalid image definition in glTF file, it should specific an 'uri' or 'bufferView'.
   at: (modules\gltf\gltf_document.cpp:3032)
ERROR: Invalid image definition in glTF file, it should specific an 'uri' or 'bufferView'.
   at: (modules\gltf\gltf_document.cpp:3032)
ERROR: Invalid image definition in glTF file, it should specific an 'uri' or 'bufferView'.
   at: (modules\gltf\gltf_document.cpp:3032)
ERROR: Index image = 0 is out of bounds (state->images.size() = 0).
   at: GLTFDocument::_get_texture (modules\gltf\gltf_document.cpp:3221)
ERROR: Index image = 0 is out of bounds (state->images.size() = 0).
   at: GLTFDocument::_get_texture (modules\gltf\gltf_document.cpp:3221)
ERROR: Index image = 1 is out of bounds (state->images.size() = 0).
   at: GLTFDocument::_get_texture (modules\gltf\gltf_document.cpp:3221)
ERROR: Index image = 1 is out of bounds (state->images.size() = 0).
   at: GLTFDocument::_get_texture (modules\gltf\gltf_document.cpp:3221)
ERROR: Index image = 2 is out of bounds (state->images.size() = 0).
   at: GLTFDocument::_get_texture (modules\gltf\gltf_document.cpp:3221)
ERROR: Index image = 3 is out of bounds (state->images.size() = 0).
   at: GLTFDocument::_get_texture (modules\gltf\gltf_document.cpp:3221)
ERROR: Index image = 0 is out of bounds (state->images.size() = 0).
   at: GLTFDocument::_get_texture (modules\gltf\gltf_document.cpp:3221)
```

Finally, I had to fix a couple lines to get the new `_import_scene` api working.

lyuma commented on editor/import/resource_importer_scene.cpp:
```diff
@@ -2131,6 +2131,24 @@ Node *EditorSceneFormatImporterESCN::import_scene(const String &p_path, uint32_t
+	for (int32_t key_i = 0; key_i < keys.size(); key_i++) {
+		options[keys[key_i]] = p_options[keys[key_i]];
+	}
+	return ResourceImporterScene::get_singleton()->import_scene_from_other_importer(this, p_path, p_flags, options, p_bake_fps);
```
```suggestion
	List<String> missing;
	Error err;
	return import_scene(p_path, p_flags, options, p_bake_fps, &missing, &err);
```

lyuma commented on editor/import/resource_importer_scene.cpp:
```diff
@@ -2131,6 +2131,24 @@ Node *EditorSceneFormatImporterESCN::import_scene(const String &p_path, uint32_t
+	for (int32_t key_i = 0; key_i < keys.size(); key_i++) {
+		options[keys[key_i]] = p_options[keys[key_i]];
+	}
+	return ResourceImporterScene::get_singleton()->import_animation_from_other_importer(this, p_path, p_flags, options, p_bake_fps);
```
```suggestion
	return import_animation(p_path, p_flags, options, p_bake_fps);
```

lyuma reviewed (commented):
I think it should call import_scene not import_scene_from_other_importer. I tried the change as is and it wasn't working, so I suggested some changes to fix this API.

I tested with my changes and `_import_scene` is working now: Here is a sample script I used to test `_import_scene` on a non-root node.
```gd
# Called when the node enters the scene tree for the first time.
func _ready():
	self.call_deferred("stupid_ready")

func set_owner_recursive(x: Node, newowner: Node):
	x.owner = newowner
	for n in x.get_children():
		set_owner_recursive(n, newowner)

func stupid_ready():
	var gltf = EditorSceneFormatImporterGLTF.new() 
	var imported = gltf._import_scene("res://test.glb.tmp", 0, {}, 30)
	imported.name = "Hello World"
	owner.add_child(imported)
	imported.owner = owner
	set_owner_recursive(imported, owner)
```

fire commented:
The scope was too much because of the addition of the importer options map api. Will revisit in a weeks time. Was recommended to split the pr into the import opens map pr and a blend api.

reduz commented:
@lyuma import_scene_from_other_importer  was removed because it was only used for importing GLTF2, which you can do with the directly exposed classes now.

lyuma commented:
I completely agree with you, @reduz 

The thing that confused me was that the API which replaces it, namely `_import_scene` starts with an underscore, and was previously only exposed to GDScript as a `virtual` call, in other words, could only be overridden but not called.

With the most recent revision, @fire bound the underscore function such that it can be called by GDScript, but that's not the usual convention I'm familiar with. Should the public API not be named `import_scene` *without* underscore?

With the most recent PR, this code works:
```gd
var fbx = EditorSceneFormatImporterFBX.new() 
var node = fbx._import_scene(fbx_path, 0, {}, 30)
```

fire commented:
Separating the change for import_scene to another pr.

fire committed:
Add support for importing .blend files

Lets you drag or place .blend files in the project folder and it will import the files.

Checks for Blender 3.0's gltf2 `export_keep_originals` option.

Add basepath support to GLTFDocument append_from_file.

Co-authored-by: Rémi Verschelde <rverschelde@gmail.com>

fire commented:
Did more testing of blender. Removed the LOD override and fixed various bugs.

I think it still breaks godot-vrm. Am still working on that.

fire commented:
Pending on https://github.com/godotengine/godot/pull/56479.

fire commented:
Test plan:

1. check every import option
2. no crashes
3. see if artists can use blender import

rcorre commented:
I just built https://github.com/godotengine/godot/pull/54886/commits/599da9d7297e0d98c5f23be60e6e6df49029961a, and I'm not seeing `filesystem/blender` under the Editor Settings. Is there something I'm missing? I'm compiling with `scons -j16 platform=x11 target=release_debug tools=yes use_llvm=no`

rcorre commented:
https://github.com/godotengine/godot/blob/599da9d7297e0d98c5f23be60e6e6df49029961a/modules/gltf_blend/register_types.cpp#L41

Hardcoding that to `true` works, but `enabled` doesn't seem to show up in ProjectSettings (the `path` config does show up in `EditorSettings` after hardcoding the former).

rcorre commented:
After opening a project a few times, it crashed with "corrupted double-linked list". I'm unsure if this was related to blend import.

```
#0  0x00007fc071a9dd22 in raise () from /usr/lib/libc.so.6
#1  0x00007fc071a87862 in abort () from /usr/lib/libc.so.6
#2  0x00007fc071adfd28 in __libc_message () from /usr/lib/libc.so.6
#3  0x00007fc071ae792a in malloc_printerr () from /usr/lib/libc.so.6
#4  0x00007fc071ae880c in unlink_chunk.constprop () from /usr/lib/libc.so.6
#5  0x00007fc071aeb27a in _int_malloc () from /usr/lib/libc.so.6
#6  0x00007fc071aec4e1 in malloc () from /usr/lib/libc.so.6
#7  0x00005633b5534383 in Memory::alloc_static (p_bytes=80, p_pad_align=<optimized out>) at core/os/memory.cpp:75
#8  0x00005633b4735f42 in Map<Vector3, LocalVector<Pair<int, int>, unsigned int, false>, Comparator<Vector3>, DefaultAllocator>::_insert (this=this@entry=0x7ffdd27137b0, p_key=..., p_value=...) at ./core/templates/map.h:411
#9  0x00005633b473643c in Map<Vector3, LocalVector<Pair<int, int>, unsigned int, false>, Comparator<Vector3>, DefaultAllocator>::insert (p_value=..., p_key=..., this=0x7ffdd27137b0) at ./core/templates/map.h:624
#10 Map<Vector3, LocalVector<Pair<int, int>, unsigned int, false>, Comparator<Vector3>, DefaultAllocator>::operator[] (this=this@entry=0x7ffdd27137b0, p_key=...) at ./core/templates/map.h:673
#11 0x00005633b47256ca in ImporterMesh::generate_lods (this=<optimized out>, p_normal_merge_angle=p_normal_merge_angle@entry=60, p_normal_split_angle=p_normal_split_angle@entry=25) at ./core/templates/local_vector.h:230
#12 0x00005633b338ea7c in ResourceImporterScene::_generate_meshes (this=0x5633bc695a00, p_node=0x5633bbf4a4a0, p_mesh_data=..., p_generate_lods=true, p_create_shadow_meshes=true,
    p_light_bake_mode=ResourceImporterScene::LIGHT_BAKE_STATIC, p_lightmap_texel_size=0.100000001, p_src_lightmap_cache=..., r_lightmap_caches=...) at ./core/object/ref_counted.h:100
#13 0x00005633b338e23c in ResourceImporterScene::_generate_meshes (this=0x5633bc695a00, p_node=0x5633c5789790, p_mesh_data=..., p_generate_lods=true, p_create_shadow_meshes=true,
    p_light_bake_mode=ResourceImporterScene::LIGHT_BAKE_STATIC, p_lightmap_texel_size=0.100000001, p_src_lightmap_cache=..., r_lightmap_caches=...) at editor/import/resource_importer_scene.cpp:1635
#14 0x00005633b338e23c in ResourceImporterScene::_generate_meshes (this=0x5633bc695a00, p_node=0x5633c6d80770, p_mesh_data=..., p_generate_lods=true, p_create_shadow_meshes=true,
    p_light_bake_mode=ResourceImporterScene::LIGHT_BAKE_STATIC, p_lightmap_texel_size=0.100000001, p_src_lightmap_cache=..., r_lightmap_caches=...) at editor/import/resource_importer_scene.cpp:1635
#15 0x00005633b339540a in ResourceImporterScene::import (this=0x5633bc695a00, p_source_file=..., p_save_path=..., p_options=..., r_platform_variants=<optimized out>, r_gen_files=<optimized out>, r_metadata=0x7ffdd27142e0)
    at editor/import/resource_importer_scene.cpp:1971
#16 0x00005633b2f07b1d in EditorFileSystem::_reimport_file (this=0x5633bd4309e0, p_file=..., p_custom_options=<optimized out>, p_custom_importer=...) at editor/editor_file_system.cpp:1878
#17 0x00005633b2f0e709 in EditorFileSystem::reimport_files (this=0x5633bd4309e0, p_files=...) at editor/editor_file_system.cpp:2127
#18 0x00005633b2f105d5 in EditorFileSystem::_update_scan_actions (this=0x5633bd4309e0) at editor/editor_file_system.cpp:626
#19 0x00005633b2f1202c in EditorFileSystem::_notification (this=0x5633bd4309e0, p_what=<optimized out>) at editor/editor_file_system.cpp:1214
#20 0x00005633b59c99a4 in Object::notification (this=0x5633bd4309e0, p_notification=17, p_reversed=<optimized out>) at core/object/object.cpp:846
#21 0x00005633b3e0e812 in SceneTree::_notify_group_pause (this=<optimized out>, p_group=..., p_notification=17) at scene/main/scene_tree.cpp:855
#22 0x00005633b3e100cf in SceneTree::process (this=0x5633bac40b40, p_time=0.14964233333333332) at scene/main/scene_tree.cpp:453
#23 0x00005633b1ee7327 in Main::iteration () at main/main.cpp:2687
#24 0x00005633b1eba469 in OS_LinuxBSD::run (this=this@entry=0x7ffdd2714d60) at platform/linuxbsd/os_linuxbsd.cpp:342
#25 0x00005633b1ea9172 in main (argc=2, argv=0x7ffdd2715278) at platform/linuxbsd/godot_linuxbsd.cpp:58
```

fire commented:
The code fails at ImporterMesh::generate_lods @JFonS any ideas?

akien-mga commented on modules/gltf_blend/editor_scene_importer_blend.cpp:
```diff
@@ -0,0 +1,272 @@
+#include "core/io/json.h"
+#include "core/string/print_string.h"
+#include "editor/editor_settings.h"
+#include "modules/gltf/editor_scene_importer_gltf.h"
+#include "modules/gltf/gltf_document.h"
```
I would suggest moving this code to the `gltf` module directly, so you don't need to have cross-module dependencies like this. If it's only usable with gltf, and doesn't add a lot of code or dependencies, it can all be in the same module.

fire replied:
For code organization purposes I'd like it to be separate if possible.

fire replied:
Don't have a strong opinion. The same structure is used for FBX. There's a lot of code in registration.

akien-mga replied:
It's a single file, and it depends on the gltf module, so I still think it should go in `modules/gltf/editor/`.
The same will apply for the FBX to glTF importer. It's cleaner IMO if all glTF features are in the `gltf` module.

It can also also be in `modules/gltf/editor/blend/` if you want it in its own folder.

akien-mga replied:
Done in https://github.com/akien-mga/godot/tree/blend-io

akien-mga commented on modules/gltf_blend/editor_scene_importer_blend.cpp:
```diff
@@ -0,0 +1,277 @@
+		const String &p_path, const String &p_option,
+		const Map<StringName, Variant> &p_options) {
+	if (p_option.begins_with("animation/")) {
+		if (p_option != "animation/import" &&
+				!bool(p_options["animation/import"])) {
```
I think your IDE hard wraps a bit too aggressively. We don't have a max line length and wrapping a relatively short condition like this doesn't really make things more readable IMO.

Not a big deal though.

akien-mga commented on modules/gltf/gltf_document.cpp:
```diff
@@ -30,6 +30,7 @@
 
 #include "gltf_document.h"
 
+#include "core/object/class_db.h"
```
As you can see here the includes are grouped with `gltf_` local includes first, then general core/scene/editor/etc. headers next (absolute paths). So this should go in the next group.

akien-mga commented:
I pushed updates (for now as a separate commit) to https://github.com/akien-mga/godot/tree/blend-io to move the class in `modules/gltf/editor/` as discussed, and I did some cleanup, adding comments and improving overall code quality and settings validation.

It's still not perfect though, whether the import can be done is validated on editor init, so enabling Blend import requires first setting the Blender path, and then enabling the option. Otherwise before my changes you would enter import hell since it would fail repeatedly until you kill Godot, and now after my changes it will just not register the importer. But that's part of what @reduz is going to improve in a follow-up.

Otherwise, I tested the version check and it doesn't seem to work. Picking a Blender 2.93 binary, I did not see `Blender is 3.0 or higher is required.`, and instead it proceeded to run the script and fail (also repeatedly since Godot always retries after failing, but that's a separate issue).

fire commented:
I saw your message. I'll update the V-Sekai fork promptly.

fire commented:
Can leave the version check and the validation on editor initialization to future commits.

Thanks!

akien-mga commented:
Seems like I broke the editor with a last minute change, will fix.

akien-mga commented:
@fire I updated my branch if you can push to this PR again: https://github.com/akien-mga/godot/tree/blend-io

akien-mga commented:
Thanks!

rcorre commented:
When I start the editor, I only see "Blender Path" in EditorSettings, not "Import Enabled", and just setting "Blender Path" does not seem to enable blend imports.

Zireael07 commented:
Where do I find the docs on this? Does this support geometry nodes?

rcorre commented:
@Zireael07 it just exports to `gltf` behind-the-scenes, so feature-wise it is the same as https://docs.godotengine.org/en/3.4/tutorials/assets_pipeline/importing_scenes.html.
